src/dmi_decode.c

```diff
@@ -0,0 +1,250 @@
+    if (buf[0x06] > 0x20) {
+        fprintf(stderr,
+            "Entry point length too large (%u bytes, expected %u).\n",
+            (unsigned int)buf[0x06], 0x18U);
+        return -1;
+    }
+
+    /* From specification: obtain offset for System Information */
+    len  = (uint32_t)((buf + 0x0C)[0] + ((buf + 0x0C)[1] << 8)
+                     + ((buf + 0x0C)[2] << 16) + ((buf + 0x0C)[3] << 24));
+ 
+    dmi_buf = read_file(0, &len, devmem);
+    if (dmi_buf) {
+        uint8_t *data, *next;
+        struct dmi_header h;
+
+        data = dmi_buf;
+        while (data + 4 <= dmi_buf + len) {
+            h.type = data[0];
+            h.length = data[1];
+            h.handle = data[2] + (data[3] << 8);
+            h.data = data;
+
+            if (h.type == 1) {
+                const char *sn, *pn, *mfr;
+                uint8_t *sys_data = h.data;
+                if (h.length < 0x08)
+                    break;
+                mfr = dmi_string(&h, sys_data[0x04]);
+                pn  = dmi_string(&h, sys_data[0x05]);
+                sn  = dmi_string(&h, sys_data[0x07]);
+
+                meta->manufacturer   = malloc(strlen(mfr) + 1);
+                meta->product_name   = malloc(strlen(pn) + 1);
+                meta->serial_number  = malloc(strlen(sn) + 1);
+                if (!meta->manufacturer || !meta->product_name || !meta->serial_number) {
+                    fprintf(stderr, "Cannot allocate memory.\n");
+                    ret = -1;
+                    break;
+                }
+                strcpy(meta->manufacturer, mfr);
+                strcpy(meta->product_name, pn);
+                strcpy(meta->serial_number, sn);
+
+                printf("Manufacturer >%s<\n", meta->manufacturer);
+                printf("Product Name >%s<\n", meta->product_name);
+                printf("Serial Number >%s<\n", meta->serial_number);
+                break;
+            }
+
+            next = data + h.length;
+            while ((unsigned long)(next - dmi_buf + 1) < len && (next[0] != 0 || next[1] != 0))
+                next++;
+
+            next += 2; /* This is also some magic from dmidecode */
+            /* Make sure the whole structure fits in the table */
+            if ((unsigned long)(next - dmi_buf) > len)
+                break;
+
+            data = next;
+        }
+
+        free(dmi_buf);
+    }
+
+    return ret;
+}
+
+static inline void free_null_ptr(void *ptr)
+{
+    if (ptr != NULL) {
+        free(ptr);
+        ptr = NULL;
+    }
+}
+
+int get_dmi_system_info(struct meta_data *meta)
+{
+    size_t size;
+    void *buf = NULL;
+
+    size = 0x20;
+    if ((buf = read_file(0, &size, SYS_ENTRY_FILE)) != NULL) {
+        if (size >= 24 && memcmp(buf, "_SM3_", 5) == 0)
+            if (att_smbios3_decode(buf, SYS_TABLE_FILE, meta) < 0) {
+                free_null_ptr(meta->manufacturer);
+                free_null_ptr(meta->product_name);
+                free_null_ptr(meta->serial_number);
```

Comment:
@0xDen You forgot to return an error.
```suggestion
                free_null_ptr(meta->serial_number);
                return -1;
```

Reply:
Fixed